src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs

```diff
@@ -0,0 +1,238 @@
+        }
+        #endregion
+
+        #region TokenSerialization
+
+        /// <summary>
+        /// Indicates whether the current XML element can be read as a token 
+        /// of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at a start 
+        /// element. The reader should not be advanced.</param>
+        /// <returns>'True' if the ReadToken method can the element.</returns>
+        public override bool CanReadToken(XmlReader reader)
+        {
+            return _internalSaml2SecurityTokenHandler.CanReadToken(reader);
+        }
+
+        /// <summary>
+        /// Deserializes from XML a token of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at the token's start 
+        /// element.</param>
+        /// <returns>An instance of <see cref="Saml2SecurityToken"/>.</returns>
+        /// <exception cref="InvalidOperationException">Is thrown if 'Configuration' or 'Configruation.IssuerTokenResolver' is null.</exception>
+        public override SecurityToken ReadToken(XmlReader reader)
+        {
+            XmlDocument doc = new XmlDocument();
+            XmlElement rstXml = (doc.ReadNode(reader) as XmlElement);
+            MSIdentityTokens.Saml2.Saml2SecurityToken internalSecurityToken = (MSIdentityTokens.Saml2.Saml2SecurityToken)_internalSaml2SecurityTokenHandler.ReadToken(rstXml.OuterXml);
+            SecurityToken token;
+            MSIdentityTokens.SecurityKey securityKey;
+            TryResolveIssuerToken(internalSecurityToken.Assertion, Configuration.IssuerTokenResolver, out token, out securityKey);
+            Saml2SecurityToken saml2SecurityToken = new Saml2SecurityToken(internalSecurityToken);
+            saml2SecurityToken.SigningToken = token;
+            saml2SecurityToken.AssertionXML = rstXml.OuterXml;
+            internalSecurityToken.SigningKey = securityKey;
+            return saml2SecurityToken;
+        }
+
+        protected virtual bool TryResolveIssuerToken(Saml2Assertion assertion, SecurityTokenResolver issuerResolver, out SecurityToken token, out MSIdentityTokens.SecurityKey securityKey)
+        {
+            if (null == assertion)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(assertion));
+            }
+            X509Certificate2 finalCertificate = null;
+            foreach (var objdata in assertion.Signature.KeyInfo.X509Data)
+            {
+                foreach (String certificateStr in objdata.Certificates)
+                {
+                    byte[] data = Convert.FromBase64String(certificateStr);
+                    finalCertificate = new X509Certificate2(data);
+                    break;
+                }
+            }
+            if (finalCertificate == null)
+            {
+                throw new Exception("Certificate not found");
+            }
```

Comment:
I don't think you need to throw here. Looking at the original implementation, if assertion.SigningCredentials.SigningKeyIdentifier is null, then it sets the token to null and returns false. If the caller doesn't have an issuer token, then it needs to decide what to do instead of being thrown a random exception.